update doc to demonstrate usage of Persistence (#5356)
AzureAD/microsoft-authentication-library-for-js: extensions/docs/msal-node-extensions.md

- Linux - LibSecret is used for storing to "Secret Service" through npm keytar.

## Code

### Creating the persistence layer

API for creating the persistence layer will differ based on what platform you are targeting.

> Alternatively, you can use PersistenceCreator.createPersistence as it's a generic wrapper and selects the appropriate persistence based on the platform . 

Reviewer: ```suggestion
Alternatively, you can use the `createPersistence` API from [PersistenceCreator](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/msal-node-extensions/src/persistence/PersistenceCreator.ts) as it's a generic wrapper and selects the appropriate persistence method based on the platform/OS. 
```

```js
const { protectedResources, msalConfig, persistenceConfiguration } = require("./authConfig");
const {
  DataProtectionScope,
  PersistenceCreator,
  PersistenceCachePlugin,
} = require("@azure/msal-node-extensions");

let authProvider;
authProvider = new AuthProvider(msalConfig);
...

PersistenceCreator
        .createPersistence(persistenceConfiguration)

>         .then((persistence) => {

Reviewer 2: Should we change this to: `const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration)`?

Should we be using await instead of .then here and everywhere else in this file?

Reviewer: +1

> const { protectedResources, msalConfig, persistenceConfiguration } = require("./authConfig");
> const {
>   DataProtectionScope,
>   PersistenceCreator,
>   PersistenceCachePlugin,
> } = require("@azure/msal-node-extensions");
> 
> let authProvider;
> authProvider = new AuthProvider(msalConfig);
> ...
> 
> PersistenceCreator
>         .createPersistence(persistenceConfiguration)
>         .then((persistence) => {
>             msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);
> 
>             authProvider = new AuthProvider(msalConfig); //pass the msalConfig to AuthProvider constructor
>         });

Reviewer: ```suggestion
const { PublicClientApplication } = require("@azure/msal-node");

const {
  DataProtectionScope,
  PersistenceCreator,
  PersistenceCachePlugin,
} = require("@azure/msal-node-extensions");

PersistenceCreator
        .createPersistence({
                cachePath: "path/to/cache/file.json",
                dataProtectionScope: DataProtectionScope.CurrentUser,
                serviceName: "test-msal-electron-service",
                accountName: "test-msal-electron-account",
                usePlaintextFileOnLinux: false,
          }).then((persistence) => {
          
            // use the persistence object to initalize an 
            // MSAL PublicClientApplication with cachePlugin
            const pca = new PublicClientApplication({
                    auth: {
                        clientId: "CLIENT_ID_HERE",
                    },
                    cache: {
                        cachePlugin: new PersistenceCachePlugin(persistence);
                    },
                });
        });
```

Reviewer: wanted to make this snippet more self contained (and generic), but feel free to edit! (I think we won't need the last snippet down below (authConfig.js) after this (actually the entire last section can go away)

```

> Alternatively, you can use below platform-specific options:-

Reviewer: ```suggestion
Alternatively, you can use below platform-specific options:
```

#### Windows:
```js
import { FilePersistenceWithDataProtection, DataProtectionScope } from "@azure/msal-node-extensions";

const cachePath = "path/to/cache/file.json";
const dataProtectionScope = DataProtectionScope.CurrentUser;

> const optionalEntropy = "";

Reviewer: > const optionalEntropy = "";

Can we add a simple comment about what this does?

- Linux - LibSecret is used for storing to "Secret Service" through npm keytar.

## Code

### Creating the persistence layer

API for creating the persistence layer will differ based on what platform you are targeting.

Alternatively, you can use the `createPersistence` API from [PersistenceCreator](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/msal-node-extensions/src/persistence/PersistenceCreator.ts) as it's a generic wrapper and selects the appropriate persistence method based on the platform/OS. 

```js
const { PublicClientApplication } = require("@azure/msal-node");
const {
  DataProtectionScope,
  PersistenceCreator,
  PersistenceCachePlugin,
} = require("@azure/msal-node-extensions");

const persistence = await PersistenceCreator.createPersistence({
                cachePath: "path/to/cache/file.json",
                dataProtectionScope: DataProtectionScope.CurrentUser,
                serviceName: "test-msal-electron-service",
                accountName: "test-msal-electron-account",
                usePlaintextFileOnLinux: false,
          });

> // Pass the persistence to msal config's cachePlugin
> msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);

Reviewer: ```suggestion
```

Reviewer: these lines are superfluous

// Use the persistence object to initialize an MSAL PublicClientApplication with cachePlugin
const pca = new PublicClientApplication({
                auth: {
                        clientId: "CLIENT_ID_HERE",
                    },
                cache: {
                        cachePlugin: new PersistenceCachePlugin(persistence);
                    },
                });

> ```

Reviewer 2: Did you add this much indentation on purpose? I'm wondering why you indented everything after the first line so much.

Author: this is similar to the way we are using the config in the sample and for readability.

Alternatively, you can use below platform-specific options:
#### Windows:
```js

> const { FilePersistenceWithDataProtection, DataProtectionScope } = require("@azure/msal-node-extensions");

Reviewer 2: Just curios: why did you change from import to require?

Author: to make it consistent with js electron sample

const { PublicClientApplication } = require("@azure/msal-node");

const cachePath = "path/to/cache/file.json";
const dataProtectionScope = DataProtectionScope.CurrentUser;
const optionalEntropy = ""; //specifies password or other additional entropy used to encrypt the data.
const windowsPersistence = await FilePersistenceWithDataProtection.create(cachePath, dataProtectionScope, optionalEntropy);

> // Pass the persistence to msal config's cachePlugin
> msalConfig.cache.cachePlugin = new PersistenceCachePlugin(windowsPersistence);

Reviewer: ```suggestion
```

const cachePath = "path/to/cache/file.json";

> const serviceName = "";

Reviewer: Can we add example serviceName and accountName here, instead of empty strings?

#### Mac:
```js
const { KeychainPersistence } = require("@azure/msal-node-extensions");

const cachePath = "path/to/cache/file.json";
const serviceName = "test-msal-electron-service";
const accountName = "test-msal-electron-account";
const macPersistence = await KeychainPersistence.create(cachePath, serviceName, accountName);

> // Pass the persistence to msal config's cachePlugin
> msalConfig.cache.cachePlugin = new PersistenceCachePlugin(macPersistence);

Reviewer: ```suggestion
```

Author: addressed

const cachePath = "path/to/cache/file.json";

> const serviceName = "";

Reviewer: Can we add example `serviceName `and `accountName `here, instead of empty strings?

#### Linux:
```js
const { LibSecretPersistence } = require("@azure/msal-node-extensions");

const cachePath = "path/to/cache/file.json";
const serviceName = "test-msal-electron-service";
const accountName = "test-msal-electron-account";
const linuxPersistence = await LibSecretPersistence.create(cachePath, serviceName, accountName);

> // Pass the persistence to msal config's cachePlugin
> msalConfig.cache.cachePlugin = new PersistenceCachePlugin(linuxPersistence);

Reviewer: ```suggestion
```

```js
import { FilePersistence } from "@azure/msal-node-extensions";

const filePath = "path/to/cache/file.json";
const filePersistence = FilePersistence.create(filePath, loggerOptions).then((persistence) => {
            msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);

            authProvider = new AuthProvider(msalConfig); //pass the msalConfig to AuthProvider constructor
        });

```

> Note:- If file or directory has not been created, it FilePersistence.create() will create file and any directories in the path recursively.

Reviewer: ```suggestion
> :information_source: If file or directory has not been created, `FilePersistence.create()` will create the file and any directories in the path recursively.
```

An unencrypted file persistence, which works across all platforms, is provided for convenience, although not recommended.

```js
const { FilePersistence } = require("@azure/msal-node-extensions");

const filePath = "path/to/cache/file.json";
const filePersistence = await FilePersistence.create(filePath, loggerOptions);
// Pass the persistence to msal config's cachePlugin

> msalConfig.cache.cachePlugin = new PersistenceCachePlugin(filePersistence);
>  // Initialize the electron authenticator
> authProvider = new AuthProvider(msalConfig); 

Reviewer: ```suggestion
const pca = new PublicClientApplication({
    auth: {
            clientId: "CLIENT_ID_HERE",
        },
    cache: {
            cachePlugin: new PersistenceCachePlugin(filePersistence);
        },
  });
```

Reviewer: The reader doesn't know about our custom class `AuthProvider`, we should just instantiate `PublicClientApplication`

```

> :https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/e7ea9fd970c035aaee9f8d3c3d3196334fd1c0de/extensions/msal-node-extensions/src/persistence/FilePersistence.ts#L18: If file or directory has not been created, `FilePersistence.create()` will create the file and any directories in the path recursively.

> ### Creating the cache plugin

Reviewer: > ### Creating the cache plugin

Can we retitle this section? We already covered how to create the cachePLuing in the snippets above, the crucial piece of information in this section is the part about this:

> To support concurrent access by multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.

Create the PersistenceCachePlugin, by passing in the persistence object that was created in the previous step.

```js
const { PersistenceCachePlugin } = require("@azure/msal-node-extensions");

const persistenceCachePlugin = new PersistenceCachePlugin(windowsPersistence); // or any of the other ones.
```

> To support concurrent access my multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.

Reviewer: ```suggestion
To support concurrent access by multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.
```

```js
const {

>   PersistenceCreator

Reviewer 2: nit: missing comma at the end of this line

  PersistenceCachePlugin,
} = require("@azure/msal-node-extensions");

const lockOptions = {
    retryNumber: 100,
    retryDelay: 50
}

const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration);
const persistenceCachePlugin = new PersistenceCachePlugin(persistence, lockOptions); // or any of the other ones

> // Pass the persistence to msal config's cachePlugin
> msalConfig.cache.cachePlugin = persistenceCachePlugin;
> // Initialize the electron authenticator
> authProvider = new AuthProvider(msalConfig);

Reviewer: ```suggestion
const pca = new PublicClientApplication({
    auth: {
            clientId: "CLIENT_ID_HERE",
        },
    cache: {
            cachePlugin: persistenceCachePlugin
        },
    });
```

```

> ### Setting the PersistenceCachePlugin on the MSAL Node PublicClientApplication configuration

Reviewer: @bmahall let's remove this entire section (i.e. **###Setting the PersistenceCachePlugin on the MSAL Node PublicClientApplication configuration**) and everything under it, since in the snippets above we already show how to create a PublicClientApplication instance with the persistenceCachePlugin. For more details, we can just direct people to check the samples

Author: I have updated the title , would like to keep this for the sake of completeness.  However, if absolutely redundant happy to have a separate PR to remove it. @derisen

    retryNumber: 100,
    retryDelay: 50
}

  PersistenceCreator
        .createPersistence(persistenceConfiguration)
        .then((persistence) => {
            const persistenceCachePlugin = new PersistenceCachePlugin(windowsPersistence, lockOptions); // or any of the other ones
            msalConfig.cache.cachePlugin = persistenceCachePlugin; //set cachePlugin in msalConfig

            // Initialize the electron authenticator
            authProvider = new AuthProvider(msalConfig);
        });

```

### Setting the PersistenceCachePlugin on the MSAL Node PublicClientApplication configuration

> Once you have a PersistenceCachePlugin, that can be set on the MSAL Node PublicClientApplication, by setting it as part of the configuration .

Reviewer: ```suggestion
Once you have a `PersistenceCachePlugin`, that can be set on the MSAL Node `PublicClientApplication`, by setting it as part of the [configuration](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#configuration) object.
```

const REDIRECT_URI = "ENTER_REDIRECT_URI";

const cachePath = "path/to/cache/file.json";

/*define persistence config based on the appropriate persistence you are using(e.g- FilePersistenceWithDataProtection, generic PersistenceCreateor, etc)*/

//defining persistence config for PersistenceCreator
const persistenceConfiguration = {
    cachePath,
    dataProtectionScope: DataProtectionScope.CurrentUser,
    serviceName: "test-msal-electron-service",
    accountName: "test-msal-electron-account",
    usePlaintextFileOnLinux: false,
}

  const msalConfig = {
    auth: {
        clientId: "CLIENT_ID_HERE",
        authority: `${AAD_ENDPOINT_HOST}TENANT_ID_HERE`,
    },
    cache: {
        cachePlugin: null // set later in main.js as shown above 
    },
    system: {
        loggerOptions: {
            loggerCallback(loglevel, message, containsPii) {
                console.log(message);
            },
            piiLoggingEnabled: false,
            logLevel: LogLevel.Verbose,
        },
    },
};
...

module.exports = {
  msalConfig: msalConfig,
  protectedResources: protectedResources,
  REDIRECT_URI: REDIRECT_URI,
  persistenceConfiguration
};

```

> Note that MSAL will not read and write to persistence by default. You will have to call PublicClientApplication.tokenCache.readFromPersistence() and PublicClientApplication.tokenCache.writeToPersistence() anytime you trigger and MSAL Node operation that alters the token cache.

Reviewer: > Note that MSAL will not read and write to persistence by default. You will have to call PublicClientApplication.tokenCache.readFromPersistence() and PublicClientApplication.tokenCache.writeToPersistence() anytime you trigger and MSAL Node operation that alters the token cache.

Is this true? cc @sameerag 
I think cache read/write will be triggered whenever a public cache API is used e.g. `PublicClientApplication.tokenCache.getAllAccounts()` (not quite sure about acquireTokenSilent though)
Or am I missing the intent here?

Reviewer 3: Not true. I think it should get triggered when we set `allowPeristence: true` in config and enable the wrapper on initial pca construction. Example [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/samples/msal-node-extensions/index.js).

Author: removed

### Note for Electron Developers:
Electron sample  link will be added here soon for clarity. 

> 

Author: @derisen Thanks for your help in going over the documentation gaps and creating a separate work item to add the sample here ! Appreciate further feedback. 
CC: @salman90 @sameerag

Reviewer: Let's place [this link](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/samples/electron-webpack) now, since we will eventually update it.

Reviewer: Oh there's [this sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/samples/msal-node-extensions) too -let's link both of them in this doc

Author: This https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/samples/electron-webpack is fairly self-sufficient, the second link seems related to this one. Should I add both here? @derisen

Reviewer: Wouldn't hurt I think (usually I like to end docs with a #samples section listing all the related samples)

const pca = new PublicClientApplication(publicClientConfig);
```

> Example (for Electron node-js desktop app):-
> 
> authConfig.js:-
> ```js
> const AAD_ENDPOINT_HOST = "https://login.microsoftonline.com/"; // include the trailing slash
> const REDIRECT_URI = "ENTER_REDIRECT_URI";
> 
> const cachePath = "path/to/cache/file.json";
> 
> /*define persistence config based on the appropriate persistence you are using(e.g- FilePersistenceWithDataProtection, generic PersistenceCreateor, etc)*/
> 
> //defining persistence config for PersistenceCreator
> const persistenceConfiguration = {
>     cachePath,
>     dataProtectionScope: DataProtectionScope.CurrentUser,
>     serviceName: "test-msal-electron-service",
>     accountName: "test-msal-electron-account",
>     usePlaintextFileOnLinux: false,
> }
> 
>   const msalConfig = {
>     auth: {
>         clientId: "CLIENT_ID_HERE",
>         authority: `${AAD_ENDPOINT_HOST}TENANT_ID_HERE`,
>     },
>     cache: {
>         cachePlugin: null // set later in main.js as shown above 
>     },
>     system: {
>         loggerOptions: {
>             loggerCallback(loglevel, message, containsPii) {
>                 console.log(message);
>             },
>             piiLoggingEnabled: false,
>             logLevel: LogLevel.Verbose,
>         },
>     },
> };
> ...
> 
> module.exports = {
>   msalConfig: msalConfig,
>   protectedResources: protectedResources,
>   REDIRECT_URI: REDIRECT_URI,
>   persistenceConfiguration
> };
> 
> ```

Reviewer: ```suggestion
```
